sdk/src/azure/platform/az_pahoasync.c

```diff
@@ -0,0 +1,619 @@
+                             .topic = az_span_create((uint8_t*)topicName, topicLen),
+                             .properties = &property_bag });
+
+  if (az_result_failed(ret))
+  {
+    _az_pahoasync_critical_error();
+  }
+
+  MQTTAsync_freeMessage(&m);
+  MQTTAsync_free(topicName);
+  return 1;
+}
+
+static void _az_pahoasync_on_connect_success(void* context, MQTTAsync_successData5* response)
+{
+  az_result ret;
+  az_mqtt5* me = (az_mqtt5*)context;
+
+  ret = az_mqtt5_inbound_connack(
+      me,
+      &(az_mqtt5_connack_data){ .connack_reason = response->reasonCode,
+                                .tls_authentication_error = false });
+
+  if (az_result_failed(ret))
+  {
+    _az_pahoasync_critical_error();
+  }
+}
+
+static void _az_pahoasync_on_connect_failure(void* context, MQTTAsync_failureData5* response)
+{
+  az_result ret;
+  az_mqtt5* me = (az_mqtt5*)context;
+
+  /* If the connection fails for any reason, we don't want to keep on
+   * retrying in this example, so disconnect. Without this, the client
+   * will attempt to reconnect. */
+  ret = MQTTAsync_disconnect(*me->_internal.pahoasync_handle, NULL);
+
+  if (ret != MQTTASYNC_SUCCESS && ret != MQTTASYNC_DISCONNECTED)
+  {
+    _az_pahoasync_critical_error();
+  }
+
+  // TODO_L: Failure to connect returns a successful return code in response->reasonCode.
+  // Investigation needed to determine if this is a bug in Paho.
+  ret = az_mqtt5_inbound_connack(
+      me,
+      &(az_mqtt5_connack_data){ .connack_reason = response->code,
+                                .tls_authentication_error = false });
+
+  if (az_result_failed(ret))
+  {
+    _az_pahoasync_critical_error();
+  }
+}
+
+static void _az_pahoasync_on_subscribe_success(void* context, MQTTAsync_successData5* response)
+{
+  az_mqtt5* me = (az_mqtt5*)context;
+
+  az_result ret = az_mqtt5_inbound_suback(me, &(az_mqtt5_suback_data){ .id = response->token });
+
+  if (az_result_failed(ret))
+  {
+    _az_pahoasync_critical_error();
+  }
+}
+
+static void _az_pahoasync_on_subscribe_failure(void* context, MQTTAsync_failureData5* response)
+{
+  (void)context;
+  (void)response;
+
+  // TODO_L: Implement logic to handle failing to subscribe.
+  _az_pahoasync_critical_error();
+}
+
+static void _az_pahoasync_on_publish_success(void* context, MQTTAsync_successData5* response)
+{
+  az_mqtt5* me = (az_mqtt5*)context;
+
+  az_result ret = az_mqtt5_inbound_puback(me, &(az_mqtt5_puback_data){ .id = response->token });
+
+  if (az_result_failed(ret))
+  {
+    _az_pahoasync_critical_error();
+  }
+}
+
+static void _az_pahoasync_on_publish_failure(void* context, MQTTAsync_failureData5* response)
+{
+  (void)context;
+  (void)response;
+
+  // TODO_L: Implement logic to handle failing to publish.
+  _az_pahoasync_critical_error();
+}
+
+static void _az_pahoasync_on_disconnect_success(void* context, MQTTAsync_successData5* response)
+{
+  az_mqtt5* me = (az_mqtt5*)context;
```

Comment:
Is this called in addition to _az_pahoasync_on_disconnect? If so, we probably shouldn't send the inbound event twice

Reply:
This covers the case where we send a disconnect ourselves, `_az_pahoasync_on_disconnect` covers the case where the broker sends a disconnect to the client